### DISCUSSION TITLE
JuiceFS report OOM command not allowed

### DISCUSSION THREAD

**Discussion by wutz** [juicedata/juicefs#474] [Q&A] [answered]
Today, i noticed that our juicefs cannot accessed with error `input/output error`.

So i checked juicefs logs and found that it's output `juicefs[59] <ERROR>: error: OOM command not allowed when used memory > 'maxmemory'`.

This is a Redis error, I then checked Aliyun Redis and found that the maximum memory reached 1G.

```sh
$ df -h /jfs
Filesystem       Size  Used Avail Use% Mounted on
JuiceFS:juicefs  1.0P  302G  1.0P   1% /jfs

$ df -i /jfs/
Filesystem        Inodes IUsed    IFree IUse% Mounted on
JuiceFS:juicefs 10495179  9419 10485760    1% /jfs
```

Only 9149 inodes consume 1 GB of Redis memory, is this normal ?

How many files need to be configured with a reference for Redis memory ?

> We use juicefs as prometheus and fluentd backend store.

**Comment by xiaogaozi** [comment, marked as answer]
Looks like 115 KiB per inode, could you restart Redis to observe if the memory load will down?

**Reply by xiaogaozi** [reply]
Or run `juicefs gc` to compact small objects into bigger ones.

**Reply by wutz** [reply]
I increase redis memory to 2 G and restart, it's still uses 1 G. 

I will be to try `juicefs gc`.

**Reply by wutz** [reply]
Today I execute `juicefs gc` without any arguments, it's output too many `juicefs[286599] <WARNING>: delete chunk 5501191 (9518 bytes): chunk not found`.

I used Redis command `memory stats` to observe usage and found that the `total.allocated` had gone from 1GB to 600MB (`keys.count` from 1200w to 670w) and was still going on.

**Reply by xiaogaozi** [reply]
What's your JuiceFS version (`juicefs version`)?

**Reply by wutz** [reply]
I use `juicefs-csi-driver: v0.7.0` with `juicefs: v0.10.0`, and execute `juicefs gc: v0.12.1`

**Reply by wutz** [reply]
I enable trace level log, it's output:

```
2021/06/01 17:54:47.253505 juicefs[516034] <INFO>: Meta address: redis://:Kb8E930741AA35f0@r-2zeyqytvdfqoqh4kxl.redis.rds.aliyuncs.com:6379/1
2021/06/01 17:54:47.255509 juicefs[516034] <WARNING>: maxmemory_policy is "volatile-lru", please set it to 'noeviction'.
2021/06/01 17:54:47.255758 juicefs[516034] <INFO>: Ping redis: 215.234µs
2021/06/01 17:54:47.256093 juicefs[516034] <DEBUG>: Creating oss storage at endpoint https://pt-oss-juicefs-stg.oss-cn-beijing-internal.aliyuncs.com
2021/06/01 17:54:47.258909 juicefs[516034] <INFO>: Data use oss://pt-oss-juicefs-stg/juicefs/
2021/06/01 17:54:47.258933 juicefs[516034] <INFO>: Cache: memory capacity: 300 MB
2021/06/01 17:54:47.258983 juicefs[516034] <DEBUG>: Iterating objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ start ""
2021/06/01 17:54:47.259026 juicefs[516034] <DEBUG>: Listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker ""
2021/06/01 17:54:47.363545 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 104.540736ms
2021/06/01 17:54:47.363724 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/218_5_4194304"
2021/06/01 17:54:47.429850 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 66.14568ms
2021/06/01 17:54:47.429959 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/334_0_4194304"
2021/06/01 17:54:47.491266 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 61.296915ms
2021/06/01 17:54:47.491357 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/489_0_4194304"
2021/06/01 17:54:47.549914 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 58.548486ms
2021/06/01 17:54:47.550002 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/603_4_4194304"
2021/06/01 17:54:47.615896 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 65.880142ms
2021/06/01 17:54:47.616081 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/715_9_4194304"
2021/06/01 17:54:47.672814 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 56.719539ms
2021/06/01 17:54:47.672895 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/832_0_4194304"
2021/06/01 17:54:47.742441 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 69.537394ms
2021/06/01 17:54:47.742535 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/0/946_15_4194304"
2021/06/01 17:54:47.807757 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 65.208523ms
2021/06/01 17:54:47.807850 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/1/1062_6_4194304"
2021/06/01 17:54:47.864967 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 57.103107ms
2021/06/01 17:54:47.865067 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/1/1188_12_4194304"
2021/06/01 17:54:47.923579 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 58.5025ms
2021/06/01 17:54:47.923670 juicefs[516034] <DEBUG>: Continue listing objects from oss://pt-oss-juicefs-stg/juicefs//chunks/ marker "0/1/1310_4_4194304"
2021/06/01 17:54:47.985145 juicefs[516034] <DEBUG>: Found 1000 object from oss://pt-oss-juicefs-stg/juicefs//chunks/ in 61.463638ms
2021/06/01 17:54:56.299895 juicefs[516034] <DEBUG>: DELETE chunks/14/14694/14694093_0_36663 (oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=60B603F0F31FA3383579E76D, 0.010s)
2021/06/01 17:54:56.299912 juicefs[516034] <WARNING>: delete chunk 14694093 (36663 bytes): chunk not found
2021/06/01 17:54:56.318342 juicefs[516034] <DEBUG>: DELETE chunks/14/14693/14693964_0_807 (oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=60B603F0F31FA33835DAE76D, 0.003s)
2021/06/01 17:54:56.318359 juicefs[516034] <WARNING>: delete chunk 14693964 (807 bytes): chunk not found
2021/06/01 17:54:56.323893 juicefs[516034] <DEBUG>: DELETE chunks/14/14694/14694102_0_222 (oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=60B603F0F31FA33835EAE76D, 0.003s)
2021/06/01 17:54:56.323913 juicefs[516034] <WARNING>: delete chunk 14694102 (222 bytes): chunk not found
2021/06/01 17:54:56.355327 juicefs[516034] <DEBUG>: DELETE chunks/14/14693/14693897_0_238 (oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=60B603F0F31FA3383573E86D, 0.004s)
2021/06/01 17:54:56.355344 juicefs[516034] <WARNING>: delete chunk 14693897 (238 bytes): chunk not found
```

It's report: `oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.`

**Reply by wutz** [reply]
I assign full access to OSS，now it's output `juicefs[545971] <DEBUG>: DELETE chunks/2/2168/2168289_0_20428 (<nil>, 0.005s)`

**Reply by xiaogaozi** [reply]
JuiceFS v0.10.0 is too old, it's recommended upgrade juicefs-csi-driver to v0.9.0.

**Reply by wutz** [reply]
Ok, i will be upgrade it.

**Reply by xiaogaozi** [reply]
I think we could make `<WARNING>: delete chunk 14694102 (222 bytes): chunk not found` this log more precisely, "chunk not found" isn't right information.

**Reply by xiaogaozi** [reply]
I've created an issue: https://github.com/juicedata/juicefs/issues/478

**Reply by xiaogaozi** [reply]
Also another related issue: https://github.com/juicedata/juicefs/issues/479

**Reply by wutz** [reply]
I have upgrade version to `v0.13.1`， and finish `juicefs gc --delete`. The below is current usage:

* fs: inode - 10029，size - 353G
* redis: keys.count - 247532, total.allocated - 61M
* oss: count - 154072, size - 45G

Next I'm going to execute `juicefs gc --compact`.

By the way, do I need to perform juicefs gc in crontab every day ？

**Reply by davies** [reply]
it's not necessary. In the early version of JuiceFS, it may have some leak caused by bugs.

The average size of inode is 35MB, so the expected number of keys in Redis is 20K, so the actual number of keys is 10X more than expected. We will investigate into it.

**Reply by wutz** [reply]
After executing `juicefs gc --compact`, the usage is:
* fs: inode - 10055, size - 355G
* redis: keys.count - 44435, total.allocated - 45M
* oss: count - 105936, size - 45G

Do i need to perform `juicefs gc --compact` regularly ?

**Reply by wutz** [reply]
Again executing `juicefs gc --compact`:
* fs: inode - 10083, size - 355G
* redis: key.count - 25616, total.allocated - 44M
* oss: count - 105211, size - 45G

**Reply by davies** [reply]
Currently, the compaction is triggered by write or read, there are chances that some files are not compacted as expected, so it's useful to run `juicefs gc --compact` regularly (for example, daily).

**Reply by davies** [reply]
> Again executing `juicefs gc --compact`:
> 
> * fs: inode - 10083, size - 355G
> * redis: key.count - 25616, total.allocated - 44M
> * oss: count - 105211, size - 45G

These number looks reasonable now.

**Reply by davies** [reply]
@wutz Here is the memory stats in my case:
```
# Memory
used_memory_human:44.85M
used_memory_rss_human:232.88M
used_memory_peak_human:705.21M
used_memory_peak_perc:6.36%
used_memory_overhead:41028030
used_memory_startup:4995800
used_memory_dataset:5999498
used_memory_dataset_perc:14.27%
```
The actual memory used by JuiceFS is only 6MB (used_memory_dataset), most of the used_memory_human are used_memory_overhead (41MB)

**Reply by wutz** [reply]
Me too, Redis overhead is too big.

    used_memory_human:46.48M
    used_memory_dataset:5858728
    used_memory_dataset_perc:13.30%

**Reply by wutz** [reply]
Another question, is the total amount of filesystem inodes (same as total size) a hard limit ? 

The current inode total  is about 10m：

```sh
$  df -i /jfs
Filesystem        Inodes IUsed    IFree IUse% Mounted on
JuiceFS:juicefs 10496053 10293 10485760    1% /jfs
```

We are considering juicefs for [HPC](https://en.wikipedia.org/wiki/Supercomputer) scenarios, where it typically uses inode levels of 10 billion and size levels of 10 PiB. Is there a good suggestion?

**Reply by davies** [reply]
No, there is no hard limit. Redis can't handle 10 billions of files, it will need 300+GB of memory. A SQL database can support 10 billions of files, but slower than Redis, may not be fast enough for HPC scenarios.

We plan to support TiKV as meta engine, that will be a better choice than Redis and SQL database.
